### PR TITLE
update tech debt burndown report to include unused exports

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Unit tests (`npm test`) were run locally and passed
- [ ] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [ ] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

GitHub Issue Number: N/A

## What is the new behavior?

This adds a report to our CI check which prints out a little bit of info
about the number of unused exports in the code, so that we'll get around
to removing them!

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

Just a CI change.

### CODE REVIEW

File: .github/workflows/tech-debt-burndown.yml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+      - name: Checkout PR branch
+        uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.4.0
+        if: ${{ matrix.branch == 'pr' }}
+
+      - name: Install ts-prune
+        run: npm install ts-prune@0.10.3
+
+      - name: Run ESLint and write output to disk
+        run: npx ts-prune > unused-exports-${{ matrix.branch }}.txt
```
Comment (rwaskiewicz) [inline .github/workflows/tech-debt-burndown.yml:70]: Is the name of this step still accurate?

File: .github/workflows/tech-debt-burndown.yml [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.4.0
+
+      - name: Get Core Dependencies
+        uses: ./.github/workflows/actions/get-core-dependencies
+
+      - name: Download null errors file for main branch
+        uses: actions/download-artifact@fb598a63ae348fa914e94cd0ff38f362e927b741 # v3.0.0
+        with:
+          name: null_errors_main
+
```
Comment (rwaskiewicz) [inline .github/workflows/tech-debt-burndown.yml:93]: Can we add the TODO Comment to remove this when the time comes here as well?

```suggestion
        # TODO(STENCIL-446): Remove this step once `strictNullChecks` is enabled
```

(I guess-timated on the leading spacing there)

File: .github/workflows/tech-debt-burndown.yml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      - name: Download null errors file for PR
+        uses: actions/download-artifact@fb598a63ae348fa914e94cd0ff38f362e927b741 # v3.0.0
+        with:
+          name: null_errors_pr
+
+      - name: Download unused exports for main
```
Comment (rwaskiewicz) [inline .github/workflows/tech-debt-burndown.yml:99]: Can we add the TODO Comment to remove this when the time comes here as well?

```suggestion
        # TODO(STENCIL-454): Remove or change this up once we've eliminated unused exports
```

(I guess-timated on the leading spacing there)

File: .gitignore [hunk 4]
```diff
@@ -50,3 +50,4 @@ coverage/**
 
 # TODO(STENCIL-446): Remove these once `strictNullChecks` is enabled
 null_errors*.json
+unused-exports*.txt
```
Comment (rwaskiewicz) [inline .gitignore:53]:
```suggestion
# TODO(STENCIL-454): Remove or change this up once we've eliminated unused exports
unused-exports*.txt
```

File: scripts/tech-debt-burndown-report.ts [hunk 5]
```diff
@@ -1,15 +1,15 @@
 /**
- * TODO(STENCIL-446): Remove this script once `strictNullChecks` is enabled
+ * A script for formatting a Markdown report for CI on some quantities we're tracking as we work
+ * on tech debt on the current branch vs on main. The report includes some info about scritNullChecks errors,
```
Comment (rwaskiewicz) [inline scripts/tech-debt-burndown-report.ts:3]:
```suggestion
 * on tech debt on the current branch vs on main. The report includes some info about strictNullChecks errors,
```

File: scripts/tech-debt-burndown-report.ts [hunk 6]
```diff
@@ -115,6 +115,47 @@ const fileErrorCounts = countArrayEntries(prData.map((error) => error.value.path
 
 const errorCodeCounts = countArrayEntries(prData.map((error) => error.value.tsError.value.errorString));
 
+//// unused exports check ////
+// TODO(STENCIL-454): Remove or change up this report once we've eliminated unused exports
+
+/**
+ * ts-prune includes this string on lines in its output corresponding to items
+ * which are exported and not imported anywhere but which which *are* used in their
```
Comment (rwaskiewicz) [inline scripts/tech-debt-burndown-report.ts:123]:
```suggestion
 * which are exported and not imported anywhere but which *are* used in their
```

File: scripts/tech-debt-burndown-report.ts [hunk 7]
```diff
@@ -115,6 +115,47 @@ const fileErrorCounts = countArrayEntries(prData.map((error) => error.value.path
...
+ */
+const USED_IN_MODULE = '(used in module)';
+
+/**
+ * Load in data from the unused exports reports
+ */
+const unusedExportsMain = String(fs.readFileSync('./unused-exports-main.txt'));
+const unusedExportsPR = String(fs.readFileSync('./unused-exports-pr.txt'));
+
+function processUnusedExports(rawFileContents: string) {
```
Comment (rwaskiewicz) [inline scripts/tech-debt-burndown-report.ts:134]: Minor - can we JSDoc this function, and declare its return type in the function sig?

File: scripts/tech-debt-burndown-report.ts [hunk 8]
```diff
@@ -115,6 +115,47 @@ const fileErrorCounts = countArrayEntries(prData.map((error) => error.value.path
...
+  const deadCodeLines = rawFileContents
+    .trim()
+    .split('\n')
+    .filter((line) => !line.includes(USED_IN_MODULE));
+
+  const deadCode = deadCodeLines.map((line) => {
```
Comment (rwaskiewicz) [inline scripts/tech-debt-burndown-report.ts:140]: FWIW - we could return this directly - either works though IMO

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
Functionality LGTM 👍 Pretty sure all my comments are all comment/text related

File: .github/workflows/tech-debt-burndown.yml [hunk 9]
```diff
@@ -0,0 +1,139 @@
+name: Tech Debt Burndown
+# this workflow is for reporting on various metrics for the codebase that
+# we want to pay attention to. Generally these are checks of some sort that we'll
+# want to eventually 'graduate' to full CI checks (which cause builds to fail if
+# there are any errors) once we've eliminated all the problems, but until that
+# point we run them here, separate from the main build, and write a report on our
+# progress on them to each PR.
+#
```
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply .github/workflows/tech-debt-burndown.yml:8]: oops!

File: .github/workflows/tech-debt-burndown.yml (hunk 1 above)
Comment (alicewriteswrongs) [reply .github/workflows/tech-debt-burndown.yml:70]: it is not 😅

File: scripts/tech-debt-burndown-report.ts (hunk 5 above)
Comment (alicewriteswrongs) [reply scripts/tech-debt-burndown-report.ts:3]: I should really code with `setlocal spell` on more 🤦‍♀️

**Comment by alicewriteswrongs** [comment]
@rwaskiewicz ok I think feedback should be addressed now

**Review by rwaskiewicz — APPROVED** [review]

**Pull request merged**